### DISCUSSION TITLE
AuxKernel local element node numbering _i

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#18692] [Q&A General] [answered]
Hello,
I am trying to get the equivalent of `_i`  index as in kernels but for AuxKernel. `_current_node->id()` return the global id not the current node id within the current element.
I am retrieving a variable using `coupledDofValues` so I need the index of the current local node id for it. Does using `coupledValue` return the current node for the current element and remove the need for `_i` indexing?
all variables are nodal.
Thanks in advance!

**Comment by loganharbour** [comment, marked as answer]
Can you explain a little more what you are trying to do here? Is your AuxKernel operating on a nodal or an elemental variable?

**Reply by KhaledNabilSharafeldin** [reply]
I have nodal AuxVariable that I am operating on, in `computeValue()` i'm returning `_u[qp] + (*_rho[_component])[_qp]` the way I am initializing that variable `_rho[_component] = &coupledValue("rho", _component)` 
both the variable i am operating on and rho are nodal variables, rho is a nonlinearvariable solved using a kernel.
I am making sure that this kind of updating is done ON the nodes correctly, that both _u and rho refer to the same node.

Hope that this clears it up

**Reply by loganharbour** [reply]
If the AuxKernel is nodal (meaning its `variable =` is nodal), all of the structures given with `coupledValue()[0]` and `_u[_qp == 0]` will be the nodal value. That is, the system is smart enough to know that if you've got a nodal variable, all of the coupling will also be performed in a nodal sense. The physical location of `_u[0]` and `coupledValue('some_var')[0]` should be `_q_point[0] == <the node>`

 With this, I don't believe you need `coupleDofValues()` at all. `coupledValue()` will give you the result that you want.

**Reply by rwcarlsen** [reply]
@loganharbour is correct.  Nodal aux kernels set `_qp` to zero and the `coupledValue` interface automatically handles the nodal vs elemental aspect of objects.  If you just use coupledValue moose should just do the right thing.

**Reply by KhaledNabilSharafeldin** [reply]
Thanks @loganharbour  and @rwcarlsen  for clarification, but generally speaking, if I wanted to use `coupleDofValues()` is there a simple way to index it to the current node? just curious. 
I marked it as solved since the responses solved my problem.

**Reply by loganharbour** [reply]
All kernels (whether it be a Kernel or an AuxKernel) have their evaluation loop defined based on the variable type. That is:
- If nodal, execute `computeValue()` once for each node, set `_qp = 0` and fill all data structures (variable values, points, gradients, etc) for just `_qp = 0`
- If elemental, fill data structures with data for each quadrature point, and evaluate `computeValue()` for each quadrature point, setting `_qp` appropriately.

Because of how we set things up, the "current node" in the case of a _nodal_ kernel will always have a local index of 0. That is, you should see `coupledValue('var')[0] == coupledDofValues('var')[0]` and `coupledDofValues('var').size() == 1`. Therefore, there shouldn't really be a reason to use the dof values in this context.
